presto-spark/src/main/java/com/facebook/presto/spark/planner/PrestoSparkRddFactory.java

```diff
@@ -0,0 +1,296 @@
+                String leftRemoteSourcePlanId = leftRemoteSource.getId().toString();
+                String rightRemoteSourcePlanId = rightRemoteSource.getId().toString();
+
+                JavaPairRDD<Integer, SerializedPrestoSparkPage> leftChildRdd = createRdd(leftSubPlan);
+                JavaPairRDD<Integer, SerializedPrestoSparkPage> rightChildRdd = createRdd(rightSubPlan);
+
+                PlanFragment leftFragment = leftSubPlan.getFragment();
+                PlanFragment rightFragment = rightSubPlan.getFragment();
+
+                List<PlanFragmentId> leftFragmentIds = leftRemoteSource.getSourceFragmentIds();
+                checkArgument(leftFragmentIds.size() == 1, "expected to have exactly only a single source fragment");
+                checkArgument(leftFragment.getId().equals(getOnlyElement(leftFragmentIds)));
+                List<PlanFragmentId> rightFragmentIds = rightRemoteSource.getSourceFragmentIds();
+                checkArgument(rightFragmentIds.size() == 1, "expected to have exactly only a single source fragment");
+                checkArgument(rightFragment.getId().equals(getOnlyElement(rightFragmentIds)));
+
+                // This fragment only contains remote source, thus there is no splits
+                PrestoSparkTaskDescriptor taskDescriptor = createTaskDescriptor(fragment, ImmutableList.of());
+                SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(sparkTaskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
+
+                PartitioningHandle partitioning = fragment.getPartitioning();
+                checkArgument(partitioning.equals(FIXED_HASH_DISTRIBUTION));
+
+                JavaPairRDD<Integer, SerializedPrestoSparkPage> shuffledLeftChildRdd = leftChildRdd.partitionBy(new IntegerIdentityPartitioner(hashPartitionCount));
+                JavaPairRDD<Integer, SerializedPrestoSparkPage> shuffledRightChildRdd = rightChildRdd.partitionBy(new IntegerIdentityPartitioner(hashPartitionCount));
+                return JavaPairRDD.fromJavaRDD(
+                        shuffledLeftChildRdd.zipPartitions(
+                                shuffledRightChildRdd,
+                                createTaskProcessor(executorFactoryProvider, serializedTaskDescriptor, leftRemoteSourcePlanId, rightRemoteSourcePlanId, taskStatsCollector)));
+            }
+            else {
+                throw new UnsupportedOperationException();
+            }
+        }
+
+        private static List<List<ScheduledSplit>> assignSplitsToTasks(List<ScheduledSplit> scheduledSplits, int numTasks)
+        {
+            List<List<ScheduledSplit>> assignedSplits = new ArrayList<>();
+            for (int i = 0; i < numTasks; i++) {
+                assignedSplits.add(new ArrayList<>());
+            }
+
+            for (ScheduledSplit split : scheduledSplits) {
+                int taskId = split.hashCode() % numTasks;
```

Comment:
nit: Do this:

```
                int taskId = Objects.hash(split.getPlanNodeId(), split.getSequenceId())) % numTasks;
```

As discussed in https://github.com/prestodb/presto/pull/13891 ;)